tun2socks/tun2socks.go

```diff
@@ -1,74 +1,137 @@
 package tun2socks
 
 import (
+	"errors"
+	"io"
 	"log"
 	"net"
+	"os"
+	"time"
 
 	"github.com/eycorsican/go-tun2socks/common/dns/cache"
 	"github.com/eycorsican/go-tun2socks/common/dns/fakedns"
 	"github.com/eycorsican/go-tun2socks/core"
 	"github.com/eycorsican/go-tun2socks/proxy/socks"
+	"github.com/songgao/water"
 )
 
-var lwipStack core.LWIPStack
-var isStopped = false
-
-// PacketFlow should be implemented in Java/Kotlin.
-type PacketFlow interface {
-	// WritePacket should writes packets to the TUN fd.
-	WritePacket(packet []byte)
-}
-
-// InputPacket Write IP packets to the lwIP stack. Call this function in the main loop of
-// the VpnService in Java/Kotlin, which should reads packets from the TUN fd.
-func InputPacket(data []byte) {
-	if lwipStack != nil {
-		lwipStack.Write(data)
-	}
-}
+var (
+	lwipWriter        io.Writer
+	lwipStack         core.LWIPStack
+	isStopped         bool = false
+	mtuUsed           int
+	stopSignalChannel chan bool
+	tunDev            *water.Interface
+)
 
 // Stop stop it
 func Stop() {
+	log.Printf("enter stop")
+	if stopSignalChannel != nil {
+		log.Printf("send stop sig")
+		close(stopSignalChannel)
+		time.Sleep(2 * time.Second)
```

Comment:
Why two seconds?

Reply:
Make sure before Stop returned, the goroutine is terminated